embassy-stm32/src/dma/mod.rs

```diff
@@ -29,27 +30,91 @@ pub trait Channel: sealed::Channel {
     where
         Self: 'a;
 
-    fn read<'a>(
+    type CompletionFuture<'a>: Future<Output = ()> + 'a
+    where
+        Self: 'a;
+
+    fn read_u8<'a>(
```

Comment:
the futures-based methods (`read_uX`, `write_uX`, `wait_for_completion`) can be implemented on top of the non-futures ones (`start_xxx`, `is_stopped`, `set_waker`). 

Could you move them out of the trait? Making them standalone funcs instead, like  `async fn read_u8(ch: &mut impl Channel,  ...)`. This will make the trait much smaller, simplifying the dma/bdma impls.